qcengine/programs/openmm.py

```diff
@@ -0,0 +1,182 @@
+        for atom1, atom2, bo in jmol.connectivity:
+            rw_mol.AddBond(atom1, atom2, bond_types[bo])
+
+        mol = rw_mol.GetMol()
+
+        # Write out the conformer
+        natom = len(jmol.symbols)
+        conf = Chem.Conformer(natom)
+        bohr2ang = ureg.conversion_factor("bohr", "angstrom")
+        for line in range(natom):
+            conf.SetAtomPosition(line, (bohr2ang * jmol.geometry[line, 0],
+                                        bohr2ang * jmol.geometry[line, 1],
+                                        bohr2ang * jmol.geometry[line, 2]))  # yapf: disable
+
+        mol.AddConformer(conf)
+        Chem.rdmolops.SanitizeMol(mol)
+
+        return mol
+
+    def compute(self, input_data: 'ResultInput', config: 'JobConfig') -> 'Result':
+        """
+        Runs OpenMM on given structure, inputs, in vacuum.
+        """
+        self.found(raise_error=True)
+
+        from simtk.openmm.app import Simulation
+        from simtk.openmm import units
+
+        from openforcefield.topology import Molecule
+        from openforcefield.typing.engines.smirnoff import ForceField
+        
+        ret_data["success"] = False
+
+        # Set workdir to scratch
+        # Location resolution order config.scratch_dir, /tmp
+        parent = config.scratch_directory
+        with temporary_directory(parent=parent, suffix="_openmm_scratch") as tmpdir:
+
+            # Grab molecule, forcefield
+            jmol = input_data.molecule
+            
+            # TODO: If urls are supported by
+            # `openforcefield.typing.engines.smirnoff.ForceField` already, we
+            # can eliminate the `offxml` and `url` distinction
+            # URL processing can happen there
+            if input_data.model.offxml:
+                offxml = input_data.offxml
+            elif input_data.model.url:
+                # TODO: add pull of offxml data from a url
+                pass
+            else:
+                raise InputError("OpenMM requires either `model.offxml` or `model.url` to be set")
+
+            # Process molecule with RDKit
+            rdkit_mol = self._process_molecule_rdkit(jmol)
+
+            # Create an OFF Molecule from the RDKit Molecule
+            off_mol = Molecule(rdkit_mol)
+
+            # Load an OFF Force Field
+            off_forcefield = ForceField(offxml)
+
+            # Create OpenMM system in vacuum from forcefield, molecule
+            off_top = off_mol.to_topology()
+            openmm_top = off_top.to_openmm()
+            openmm_system = off_forcefield.create_openmm_system(off_top)
+            
+            # Need an integrator for simulation even if we don't end up using it really
+            integrator = LangevinIntegrator(300*units.kelvin, 1/units.picosecond, 0.002*units.picoseconds)
+
+            # Set platform to CPU explicitly
+            platform = openmm.Platform.getPlatformByName('CPU')
```

Comment:
Can we also do something like the following to achieve this, with the ability to set the number of threads via the `JobConfig`, or do we strictly want to leave this to the parent environment variables of the process?

```
# get number of threads to use from `JobConfig; otherwise, try environment variable
nthreads = config.nthreads
if nthreads is None:
    nthreads = os.environ.get('OPENMM_CPU_THREADS')

# Set platform to CPU explicitly
platform = openmm.Platform.getPlatformByName('CPU')

# Initialize context
context = openmm.Context(openmm_system, integrator, platform)

# set number of threads
if nthreads:
    platform.setPropertyValue(context, 'Threads', nthreads)
```

Reply:
@jchodera Will this work here?

Reply:
This should work!

Reply:
I think this should work!

Reply:
The code given above will not work.  Once the Context is created, the number of threads it uses can't be changed.  Instead, you can use any of three methods to set the number of threads.

1. Set the environment variable `OPENMM_CPU_THREADS`.
2. *Before* creating the Context set the default value for the `Threads` property: `platform.setPropertyDefaultValue('Threads', '4')`
3. Provide a value for that property to the Context constructor: `context = openmm.Context(openmm_system, integrator, platform, {'Threads':'4'})`

If you don't do any of those, it defaults to the number of logical cores in your processor.

Reply:
Excellent, thanks @peastman! I've gone with option (3).